Support Lets Encrypt DNS Challenges

tyrken opened the pull request:
This PR enables various DNS Challenges for Let's Encrypt using existing functionality in the lego library.  I've also added an override for the default functionality of DNS validation, and rough logging.

Admittedly I haven't tested it much yet, but started the PR to see if there was any interest.  Hoping to use this myself on internal fire-walled servers.

errm commented:
This looks like a really good feature to have @tyrken 

I am running a number of services inside of a firewall, so would definitely be of interest to me.

ping @containous/traefik

trecloux commented:
Hi All, 
This is very interesting for people running non public services.

I forked the fork to add support for Exoscale DNS provider : https://github.com/trecloux/traefik/tree/lego-dns-exoscale
I tested my fork and was able to use traefik to get a Let's Encrypt certificate using DNS challenge.

I think that the code mapping a DNS string to a lego DNS provider should stay in lego lib in order to avoid individual support for each provider in traefik. I'm ready to submit a pull request to lego in order to support it.

I would be very happy to see this feature integrated in traefik, I'm ready to help.

Thanks @tyrken

emilevauge commented:
@trecloux great job!

>I think that the code mapping a DNS string to a lego DNS provider should stay in lego lib in order to avoid individual support for each provider in traefik. I'm ready to submit a pull request to lego in order to support it.

I definitely agree with that. This would avoid us to maintain https://github.com/containous/traefik/pull/738/files#diff-faa8f9dd50dafacfcb04a8203ce3deebR451 and would also remove a lot of dependencies. @tyrken WDYT?

trecloux commented:
The pull request has been submited to lego library : https://github.com/xenolf/lego/pull/313
And I updated my fork to include this change https://github.com/trecloux/traefik/commits/lego-dns-exoscale
Did you guys already tried to implement integration tests for Let's encrypt/ACME support ?

tyrken commented:
Yeah - did consider as copying the code from the CLI was pretty smelly, but thought a PR on another PR had little chance of quick acceptance.  Hope xenolf agrees to the extension...

emilevauge commented:
@trecloux 
>Did you guys already tried to implement integration tests for Let's encrypt/ACME support ?

This would be awesome. I tried in the past, using https://github.com/letsencrypt/boulder, but at that time, this project was not playing well with docker compose, which is what our test framework is based on.
Another issue is that they don't publish this Docker image very often :https://hub.docker.com/r/letsencrypt/boulder/tags :'(

trecloux commented:
The pull request was merged on lego side.
@tyrken @emilevauge 
How do we proceed ? Shall I create a pull request on @tyrken 's fork ?

emilevauge commented:
@trecloux yes, it seems ok :) @tyrken can you confirm you are ok to merge his PR in your fork ?

tyrken commented:
I've rebased this PR on the latest traefik master, and updated glide for the latest xenolf - @trecloux go ahead & submit a PR to my own & re-test!

Xenolf now have a feature to add OSCP stapling - sounds like a good idea, but I've left at the backwards compatible setting for now.  Open to consideration to change!

emilevauge commented:
@trecloux waiting for a new PR from you then :)

trecloux commented:
@emilevauge : I created a pull request onto @tyrken repo : https://github.com/tyrken/traefik/pull/1

emilevauge commented:
@trecloux thanks a lot :)

trecloux commented:
@emilevauge I will try to create integration tests based on the certbot integration tests : https://github.com/certbot/certbot/tree/master/tests. They are docker-compose based.
But yes, we might have to build our own boulder images ... or work from sources.

tyrken commented:
Thanks @trecloux - your changes are merged in & passing the current tests.  I merged in the latest changes - @emilevauge do you want me to squash/rebase instead in the future?

emilevauge commented:
@tyrken indeed, we use squash/rebase to keep history clean ;) 
@trecloux awesome news!

tyrken commented:
Done.

trecloux commented:
@emilevauge I will try to create integration tests based on the certbot integration tests : https://github.com/certbot/certbot/tree/master/tests. They are docker-compose based.
But yes, we might have to build our own boulder images ... or work from sources.

emilevauge commented on glide.yaml:
```diff
@@ -102,3 +102,16 @@ import:
+  - route53
+- package: github.com/ovh/go-ovh
+- package: github.com/timewasted/linode
+- package: github.com/weppos/dnsimple-go
```
Is this still needed ?

trecloux commented:
@emilevauge No ... I created an other pull request onto @tyrken fork
https://github.com/tyrken/traefik/pull/2

tyrken commented:
Thanks - squashing into one commit again.

emilevauge commented:
Ping @containous/traefik, we need another LGTM on this :)

emilevauge commented:
@tyrken sorry, needs a rebase again :'(

emilevauge commented:
@tyrken, could you [allow editing your PR by maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)? I have minor changes to make :)

tyrken commented:
I think you should have permissions already - think I did that a while back.

emilevauge commented:
@tyrken Sorry but I cannot push to your repo: `fatal: Authentication failed for 'https://github.com/tyrken/traefik.git/'` :'(
They made some changes on azure go sdk which needs to update lego to cbd5d04c891979c23c3924f198e07ce32b39d282.
Could you rebase one last time and update `glide.yaml` with:
```
- package: github.com/xenolf/lego
  version: cbd5d04c891979c23c3924f198e07ce32b39d282
```
And then `glide up`?
I will merge this one right away.
If you don't have time, let me know, I will carry this PR in another one :)

tyrken commented:
@emilevauge rebased with updated lego ref.  Not sure why you can't edit seeing as 'Allow edits from maintainers.' is ticked!

emilevauge commented:
Thanks a lot @tyrken & @trecloux on this :)

krancour commented:
This is an awesome feature! I'll start adding this to the Traefik Helm chart as soon as possible.
